File: utbot-framework/src/main/java/org/utbot/quickcheck/internal/generator/EnumGenerator.java
/*
 The MIT License

 Copyright (c) 2010-2021 Paul R. Holser, Jr.

 Permission is hereby granted, free of charge, to any person obtaining
 a copy of this software and associated documentation files (the
 "Software"), to deal in the Software without restriction, including
 without limitation the rights to use, copy, modify, merge, publish,
 distribute, sublicense, and/or sell copies of the Software, and to
 permit persons to whom the Software is furnished to do so, subject to
 the following conditions:

 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

package org.utbot.quickcheck.internal.generator;

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.plugin.api.UtModel;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.Generator;
import org.utbot.quickcheck.random.SourceOfRandomness;

import static org.utbot.external.api.UtModelFactoryKt.classIdForType;

public class EnumGenerator extends Generator<Enum> {
    private final Class<?> enumType;

    EnumGenerator(Class<?> enumType) {
        super(Enum.class);

        this.enumType = enumType;
    }

    @Override public UtModel generate(
        SourceOfRandomness random,
        GenerationStatus status) {

        Object[] values = enumType.getEnumConstants();
        int index = random.nextInt(0, values.length - 1);
        return UtModelGenerator.getUtModelConstructor().construct(values[index], classIdForType(Enum.class));
    }

    @Override public boolean canShrink(Object larger) {
        return enumType.isInstance(larger);
    }
}

File: utbot-framework/src/main/kotlin/org/utbot/quickcheck/internal/generator/EnumGenerator.kt


package org.utbot.quickcheck.internal.generator;

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.plugin.api.UtModel;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.Generator;
import org.utbot.quickcheck.random.SourceOfRandomness;

import static org.utbot.external.api.UtModelFactoryKt.classIdForType;

public class EnumGenerator extends Generator<Enum> {
    private final Class<?> enumType;

    EnumGenerator(Class<?> enumType) {
        super(Enum.class);

        this.enumType = enumType;
    }

    @Override public UtModel generate(
        SourceOfRandomness random,
        GenerationStatus status) {

        Object[] values = enumType.getEnumConstants();
        int index = random.nextInt(0, values.length - 1);
        return UtModelGenerator.getUtModelConstructor().construct(values[index], classIdForType(Enum.class));
    }

}
